packages/utils/src/object.ts

```diff
@@ -301,84 +301,83 @@ function makeSerializable<T>(value: T, key?: any): T | string {
 }
 
 /**
- * Walks an object to perform a normalization on it
+ * normalize()
  *
- * @param key of object that's walked in current iteration
- * @param value object to be walked
- * @param depth Optional number indicating how deep should walking be performed
- * @param memo Optional Memo class handling decycling
+ * - Creates a copy to prevent original input mutation
+ * - Skip non-enumerablers
+ * - Calls `toJSON` if implemented
+ * - Removes circular references
+ * - Translates non-serializeable values (undefined/NaN/Functions) to serializable format
+ * - Translates known global objects/Classes to a string representations
+ * - Takes care of Error objects serialization
+ * - Optionally limit depth of final output
  */
-// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
-export function walk(key: string, value: any, depth: number = +Infinity, memo: MemoFunc = memoBuilder()): any {
-  const [memoize, unmemoize] = memo;
+export function normalize(input: unknown, maxDepth: number = +Infinity, maxEdges: number = 10_000): any {
+  const [memoize, unmemoize] = memoBuilder();
+  let edges = 0;
+
+  function walk(key: string, value: unknown & { toJSON?: () => string }, depth: number): any {
+    // If we reach the maximum depth, serialize whatever is left
+    if (depth === 0) {
+      edges += 1;
+      return serializeValue(value);
+    }
 
-  // If we reach the maximum depth, serialize whatever is left
-  if (depth === 0) {
-    return serializeValue(value);
-  }
+    // If value implements `toJSON` method, call it and return early
+    if (typeof value?.toJSON === 'function') {
```

Comment:
unfortunately optional chaining is less bundle size efficient. 

```suggestion
    if (value != null && typeof value.toJSON === 'function') {
```

Reply:
Could we just early return if value is undefined/null?